source/h1_encoder.c

```diff
@@ -319,6 +344,309 @@ int aws_h1_encoder_start_message(
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    bool wrote_all = true;
+    wrote_all &= aws_byte_buf_write_u8(dst, ';');
+    wrote_all &= aws_byte_buf_write_from_whole_cursor(dst, chunk_extension->key);
+    wrote_all &= aws_byte_buf_write_u8(dst, '=');
+    wrote_all &= aws_byte_buf_write_from_whole_cursor(dst, chunk_extension->value);
+    return wrote_all;
+}
+
+static bool s_fill_byte_buffer(struct aws_byte_buf *dst, struct aws_byte_cursor *src) {
+    size_t write_size = aws_min_size(dst->capacity - dst->len, src->len);
+    struct aws_byte_cursor sub_cursor;
+    sub_cursor.len = write_size;
+    sub_cursor.ptr = src->ptr;
+    if (AWS_UNLIKELY(AWS_OP_SUCCESS != aws_byte_buf_append(dst, &sub_cursor))) {
+        return false;
+    }
+    aws_byte_cursor_advance(src, write_size);
+    return 0 == src->len;
+}
+
+static bool s_chunk_line_state(struct aws_byte_buf *dst, struct aws_h1_encoder *encoder) {
+    AWS_PRECONDITION(dst);
+    AWS_PRECONDITION(encoder);
+    AWS_PRECONDITION(encoder->message);
+    AWS_PRECONDITION(encoder->message->body_chunk);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    AWS_PRECONDITION(aws_byte_buf_is_valid(&encoder->message->body_chunk->chunk_line));
+    bool completely_written = s_fill_byte_buffer(dst, &encoder->message->body_chunk->chunk_line_cursor);
+    if (completely_written) {
+        aws_byte_buf_clean_up(&encoder->message->body_chunk->chunk_line);
+        return true;
+    }
+    return false;
+}
+
+static bool s_chunk_payload_state(struct aws_byte_buf *dst, struct aws_h1_encoder *encoder, int *aws_op_result) {
+    AWS_PRECONDITION(dst);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    AWS_PRECONDITION(encoder);
+    AWS_PRECONDITION(encoder->message);
+    AWS_PRECONDITION(encoder->message->body_chunk);
+    AWS_PRECONDITION(encoder->message->body_chunk->data);
+    AWS_PRECONDITION(aws_op_result);
+
+    size_t prev_len = dst->len;
+    if (AWS_UNLIKELY(aws_input_stream_read(encoder->message->body_chunk->data, dst))) {
+        ENCODER_LOGF(
+            TRACE,
+            encoder,
+            "Failed to write body stream, error %d (%s)",
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+        *aws_op_result = AWS_OP_ERR;
+        return false;
+    }
+
+    const int64_t amount_read = dst->len - prev_len;
+    encoder->progress_bytes += amount_read;
+    ENCODER_LOGF(TRACE, encoder, "Wrote %zu body bytes to message", amount_read);
+
+    struct aws_stream_status status;
+    if (AWS_OP_SUCCESS != aws_input_stream_get_status(encoder->message->body_chunk->data, &status)) {
+        ENCODER_LOGF(
+            TRACE,
+            encoder,
+            "Failed to query body stream status, error %d (%s)",
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+        *aws_op_result = AWS_OP_ERR;
+        return false;
+    }
+
+    if (status.is_end_of_stream) {
+        return write_crlf(dst);
```

Comment:
will we be ok if the stream ends, and there's just 1 byte of capacity left in the buffer?
It seems ok to me, just want to here you say it's cool too

we'll call `write_crlf()` which will do an all-or-nothing write and return false
then we'll come back into this state in the future and hopefully aws_input_stream_read() will do nothing and we'll succeed writing CRLF. right?

Reply:
Correct, if the call to s_chunk_payload_state returns false, the state of the state machine is not advanced and the output stream task will reschedule the call on a new channel message. As long as the writecrlf() transactionally succeeds in writing everything or nothing, all will be well.